yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+import org.apache.hadoop.yarn.api.records.timeline.TimelineEntity
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.deploy.history.yarn.YarnTimelineUtils._
+import org.apache.spark.deploy.history.yarn.rest.{JerseyBinding, TimelineQueryClient}
+import org.apache.spark.deploy.history.{ApplicationHistoryInfo, ApplicationHistoryProvider, HistoryServer}
+import org.apache.spark.scheduler.{ApplicationEventListener, SparkListenerBus}
+import org.apache.spark.ui.SparkUI
+import org.apache.spark.{SparkException, Logging, SecurityManager, SparkConf}
+
+/**
+ * A  History provider which reads in the history from
+ * the YARN Timeline Service.
+ *
+ * The service is a remote HTTP service, so failure modes are
+ * different from simple file IO.
+ *
+ * 1. Application listings are asynchronous, and made on a schedule, though
+ * they can be forced (and the schedule disabled).
+ * 2. The results are cached and can be retrieved with [[getApplications()]].
+ * 3. The most recent failure of any operation is stored,
+ * The [[getLastFailure()]] call will return the last exception
+ * or `None`. It is shared across threads so is primarily there for
+ * tests and basic diagnostics.
+ * 4. Listing the details of a single application in [[getAppUI()]]
+ * is synchronous and *not* cached.
+ * 5. the [[maybeCheckHealth()]] call performs a health check as the initial
+ * binding operation of this instance. This call invokes [[TimelineQueryClient.healthCheck()]]
+ * for better diagnostics on binding failures -particularly configuration problems.
+ * 6. Every REST call, synchronous or asynchronous, will invoke [[maybeCheckHealth()]] until
+ * the health check eventually succeeds.
+ * <p>
+ * If the timeline is  not enabled, the API calls used by the web UI
+ * downgrade gracefully (returning empty entries), rather than fail.
+ * 
+ *
+ * @param sparkConf configuration of the provider
+ */
+private[spark] class YarnHistoryProvider(sparkConf: SparkConf)
+  extends ApplicationHistoryProvider with Logging {
+
+  /**
+   * The configuration here is a YarnConfiguration built off the spark configuration
+   * supplied in the constructor; this operation ensures that `yarn-default.xml`
+   * and `yarn-site.xml` are pulled in. Options in the spark conf will override
+   * those in the -default and -site XML resources which are not marked as final.
+   */
+  private val yarnConf = {
+    new YarnConfiguration(SparkHadoopUtil.get.newConfiguration(sparkConf))
+  }
+
+  /**
+   * UI ACL option
+   */
+  private val uiAclsEnabled = sparkConf.getBoolean("spark.history.ui.acls.enable", false)
+
+  private val detailedInfo = sparkConf.getBoolean(YarnHistoryProvider.OPTION_DETAILED_INFO, false)
+  private val NOT_STARTED = "<Not Started>"
+
+  /* minimum interval between each check for event log updates */
+  private val refreshInterval = sparkConf.getLong(YarnHistoryProvider.OPTION_MIN_REFRESH_INTERVAL,
```

Comment:
This should probably use one of the new time APIs now (e.g. `getTimeAsSeconds`).